### PR TITLE
NewGaugeWithTags: re-check condition after acquiring write mutex

### PR DESCRIPTION
**Pull request by tflo-lyft**
was getting small number of panics on deploys in our service originating in this method: https://kibana.lyft.net/app/kibana#/doc/elasticlogs*:search-logs/elasticlogssat4:active-logs-003109/go_json?id=0caH%26kPZL%3Dsw%7D(tU%7DTP9&_g=h@6113a49

### CODE REVIEW

**Comment by tflo-lyft** [comment]
🔨

**Comment by charlievieth** [comment]
@tflo-lyft this isn't the bug, this is [`stats.go#L242`](https://github.com/lyft/gostats/blob/master/stats.go#L242).  That should be:

```golang
func (c *gauge) String() string {
	return strconv.FormatUint(atomic.LoadUint64(&c.value), 10)
}
```

File: stats.go [hunk 1]
```diff
@@ -439,11 +439,11 @@ func (s *statStore) NewGaugeWithTags(name string, tags map[string]string) Gauge
 
 	s.gaugesMtx.Lock()
 	s.gauges[name] = g
-	s.gaugesMtx.Unlock()
 
 	if s.export && expvar.Get(name) == nil {
 		expvar.Publish(name, g)
 	}
+	s.gaugesMtx.Unlock()
```
[the remark this answers is not in the document]
Comment (charlievieth) [reply stats.go]: The issue is that `expvar` can and will call `gauge.String()` whenever it wants and that method is not thread-safe.  The correct implementation of [`gauge.String()`](https://github.com/lyft/gostats/blob/bade9ed839430d28856d5b72db4fbf52fde1adfa/stats.go#L241-L243) is:

```golang
func (s *statStore) NewGaugeWithTags(name string, tags map[string]string) Gauge {
	name = serializeTags(name, tags)

	s.gaugesMtx.RLock()
	g, ok := s.gauges[name]
	s.gaugesMtx.RUnlock()
	if ok {
		return g
	}

	s.gaugesMtx.Lock()
	if g, ok := s.gauges[name]; ok {
		s.gaugesMtx.Unlock()
		return g
	}
	g = &gauge{}
	s.gauges[name] = g
	s.gaugesMtx.Unlock()

	if s.export && expvar.Get(name) == nil {
		expvar.Publish(name, g)
	}
	return g
}
```

[the remark this answers is not in the document]
Comment (charlievieth) [reply stats.go]: note that here we re-check the condition after acquiring the write lock.

**Comment by charlievieth** [comment]
👍

**Pull request merged**